### PR TITLE
Seperate mutable/immutable

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Seperate out the "mutable" version of a ProtoNode from the Immutable representation of a node tied to a block

# Implementation

Make a new MutableProtoNode class and move mutation methods over to it
Modify services that use mutations
Update tests to distinguish between mutable and immutable nodes

# For discussion

This is not so much for merging as identifying what this API surface looks like.

### CODE REVIEW

File: coding.go [hunk 1]
```diff
@@ -117,44 +95,42 @@ func (n *ProtoNode) GetPBNode() *pb.PBNode {
...
+	if err != nil {
+		return nil, err
+	}
+	return &ProtoNode{blk, upn.node}, nil
 }
 
 // DecodeProtobuf decodes raw data and returns a new Node instance.
-func DecodeProtobuf(encoded []byte) (*ProtoNode, error) {
-	n, err := unmarshal(encoded)
+func DecodeProtobuf(encoded []byte) (*UnhashedProtoNode, error) {
```
Comment (willscott) [inline coding.go:123]: where are we calling this where we don't already have the CID associated with the data?

File: dagutils/utils.go [hunk 2]
```diff
@@ -79,65 +81,85 @@ func addLink(ctx context.Context, ds ipld.DAGService, root *dag.ProtoNode, child
...
-	return root, nil
+	if err := ds.Add(ctx, nd); err != nil {
+		return nil, err
+	}
+	return nd, nil
 }
 
 // InsertNodeAtPath inserts a new node in the tree and replaces the current root with the new one.
-func (e *Editor) InsertNodeAtPath(ctx context.Context, pth string, toinsert ipld.Node, create func() *dag.ProtoNode) error {
+func (e *Editor) InsertNodeAtPath(ctx context.Context, pth string, toinsert ipld.Node, create func() *dag.MutableProtoNode) error {
```
Comment (willscott) [inline dagutils/utils.go:95]: this seems like a good candidate for the sort of thing where we may want a different interface that plays nicer with the prime view of the world.

File: coding.go (hunk 1 above)
Comment (hannahhoward) [reply coding.go:123]: yea, so partly the scary part is this is a public function, so the answer is "who knows". A cursory search using Github's organization wide search for IPFS shows several potential places.

File: dagutils/utils.go (hunk 2 above)
Comment (hannahhoward) [reply dagutils/utils.go:95]: yea, I'm talking to @warpfork about this. I beleive we want to use the FocusedTransform function to implement this, which I had forgot about before. But I think we may need to provide some enhancements withing go-ipld-prime.

**Comment by aschmahmann** [comment]
@hannahhoward what's the status of this PR?

**Pull request closed without merging**